Add autodetected http/https protocols

dblanken opened the pull request:
This is in response to @tpope's response on PR #590 to allow https to be used for localhost connections using Preview.  There, he mentioned that he'd like the uri scheme protocol to be autodetected, which I appreciate.  I opted to use 'possible approach 3' to ps/grep for ssl://<binding> to see if it was being used or not.

tpope commented:
Good start. You'll want to use `ps -p <pid>` so that it doesn't match *any* process.

Probably easiest to make `rails#get_binding_for()` responsible for this, changing the interface to return `http://localhost:3000` rather than `localhost:3000`. We already have the pid handy there, and the functions that call it (indirectly through `.server_binding()` end up prepending the protocol anyways.

dblanken commented:
Thank you for the guidance, @tpope.  ```rails#get_binding_for()``` is perfect for this.  

I didn't know how well-versed you were with the win32 side of things in ```rails#get_binding_for()```.  I attempted to install ruby/rails on a windows machine I had using RubyInstaller and was able to get it to autodetect with the code submitted, but I am not sure if that covers most of the use cases of a windows user attempting to use ```:Preview```.  It assumes that curl is present, which is available with RailsInstaller, otherwise it just reverts back to http.

The non-win32 version I think should work since I haven't seen a system without access to ps and grep, but I can add checks for that too if you think it should be there for completeness.  I was able to test this on Ubuntu and MacOS.

Also, if you'd rather the solution stick to one possible approach, I can work instead on replacing the ```ps``` solution with ```curl``` to match the win32 version, but I really like the ```ps``` solution as it seems easier to follow for me.

Thanks again for all of your help and for such a great plugin.

tpope commented:
Let's start by doing _just_ the interface change, with hard-coded `http://`. It's hard to read the diff with both changes happening at once.

> Also, if you'd rather the solution stick to one possible approach, I can work instead on replacing the `ps` solution with `curl` to match the win32 version, but I really like the `ps` solution as it seems easier to follow for me.

I'm leaning towards using the `curl` solution everywhere, as I have no idea if that `ssl:` match is the same across all web servers.

I don't actually consider Windows support mandatory for merging, but indeed an added bonus of the `curl` approach is it works everywhere.

tpope commented:
Actually hold off; the interface change is tricky and touches some bugs that need to be fixed. I'll take care of it and you can pick up from there.

tpope commented:
Okay have at it.

dblanken commented:
I tried to break up the commits a little more this time.  Also I wondered if you could verify my regex fix to make sure you're having the issue as well before the fix.  I was only able to test that on my Mac, so that's what that was based on.  I can revert if needed.

I was also iffy on if you were ok with a local method or not for this--I figured since it had to be done twice in that method, it helped not make it so long.

Again, thanks for looking at all of this and for your feedback/help.

tpope commented:
> I tried to break up the commits a little more this time. Also I wondered if you could verify my regex fix to make sure you're having the issue as well before the fix. I was only able to test that on my Mac, so that's what that was based on. I can revert if needed.

Yeah I'm not sure why I didn't catch that when testing. I've gone ahead and ported it to master.

> I was also iffy on if you were ok with a local method or not for this--I figured since it had to be done twice in that method, it helped not make it so long.

I don't see why it needs to be done twice. I deliberately restructured it to have a singular `return` at the end, with your upcoming change in mind.

dblanken commented:
I'll remove the regex hopefully tonight from the PR.  Thanks for verifying and porting that one.

For the return, I think it was [line 1991](https://github.com/tpope/vim-rails/blob/12b2f1a9c8d5394a9dc4e63267b2abad38eb1492/autoload/rails.vim#L1991) that I then broke up in [commit 1549556](https://github.com/tpope/vim-rails/pull/592/commits/1549556eccfd3dced87fc98a835de1ead1dcef13) in order to add what I consider the second check.  I did notice you gave a great avenue to one return everywhere else, but thought that possibly remained for a different reason.  I can revert that change there and just have it at the end before the return if you feel it isn't needed in that return.

tpope commented:
You can delete that `return` line, it was a mistake.

dblanken commented:
Sorry for such a delay in getting back.  I removed the line, which allowed us to put the method details back into ```get_binding_for```, and I added wget as a backup to curl if curl weren't to exist.

tpope commented on autoload/rails.vim:
```diff
@@ -1988,16 +1990,26 @@ function! rails#get_binding_for(pid) abort
       let output = system('netstat -antp')
       let binding = matchstr(output, '\S\+:\d\+\ze\s\+\S\+\s\+LISTEN\s\+'.a:pid.'/')
-      return empty(binding) ? '' : 'http://' . s:sub(binding, '^([^[]*:.*):', '[\1]:')
+      if empty(binding)
+        return ''
+      endif
```
Can be dropped; we already guard against this after the outer `endif`.

dblanken replied:
Pushing it now.  Thanks for catching that.

tpope commented:
Merged with some tweaks, thanks.

dblanken commented:
Tested your merge on my Mac.

I think the ```--max-time``` on the [curl command](https://github.com/tpope/vim-rails/blob/24c31a1d2d0ebf6fd61044deb214e07eb4f408d9/autoload/rails.vim#L2001) should not have the equals sign.
```call system('curl --max-time 2 -k --silent --head --fail ' . shellescape('https://'.binding))```

```wget``` does use the equals sign in options, so that one is ok.

Thanks for all of your guidance and help on adding this.  I really appreciate the opportunity to contribute.
